### ISSUE TITLE
Crash at the beginning of pass (ver 0.3.1)

### ISSUE THREAD

**Issue by arope99** [wose/satnogs-monitor#21]
I’m running the satnogs-monitor on an Raspberry Pi 4 with the following command as instructed:

satnogs-monitor -l 1382 --rotctld-address 127.0.0.1:4533 --data-path /tmp/.satnogs/data/ --waterfall --spectrum

However, at the start of the pass, satnogs-monitor always crashed with the following message:

Satnogs-monitor thread ‘main’ panicked at ‘called Option::unwrap() on a None value’, src/libcore/option.rs:378:21
note: run with RUST_BACKTRACE=1 environment variable to display a backtrace.

Here’s the output with RUST_BACKTRACE=1

thread ‘main’ panicked at ‘called Option::unwrap() on a None value’, src/libcore/option.rs:378:21
stack backtrace:
0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
1:
2:
3:
4:
5:
6:
7:
8:
9:
10:
11:
12:
13:
14:
15:
16:
17:
18:
19: main
20: __libc_start_main
note:
Some details are omitted, run with RUST_BACKTRACE=full for a verbose backtrace.

And, with RUST_BACKTRACE=full

thread ‘main’ panicked at ‘called Option::unwrap() on a None value’, src/libcore/option.rs:378:21
stack backtrace:
0: 0x806e3c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf24f776311cdf529
1: 0x824414 -
2: 0x800a7c -
3: 0x8091d0 - <unknown
4: 0x808e10 -
5: 0x8098d0
6: 0x8094d0 -
7: 0x8093c8 -
8: 0x820030 -
9: 0x81ff74 -
10: 0x43f5f0 -
11: 0x469c2c -
12: 0x43d058 -
13: 0x43e484 -
14: 0x45c5f4 -
15: 0x47a064 -
16: 0x809340 -
17: 0x80d6d4 -
18: 0x809e28 -
19: 0x45fc98 - main
20: 0xb6d9f718 - __libc_start_main

**Comment by wose** [comment]
Hrm, it seems like the formatting ate all the interesting bits, or are there actually no function names behind the call stack pointers (7: 0x...)?

**Comment by arope99** [comment]
I've compiled from source and run the satnogs-monitor, still with similar issues at the beginning of a pass. However, I've got better info to share when running with RUST_BACKTRACE=full.

stack backtrace:
0:   0x8077a4 - backtrace::backtrace::libunwind::trace::h100c6c70e4af496c
1:   0x8077a4 - backtrace::backtrace::trace_unsynchronized::hf78ddc194bf45009
2:   0x8077a4 - std::sys_common::backtrace_print_fmt::h967a212a8894e4a4
3:   0x8077a4 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h5fe383ce2118859a at src/libstd/sys_common/backtrace.rs:59
4:   0x823384 - core::fmt::write::h937f59172f6eeeaf at src/libcore/fmt/mod.rs:1052
5:   0x800dc8 - std::io::Write::write_fmt::hc70e2634948b89c8 at src/libstd/io/mod.rs:1426
6:   0x809c1c - std::sys_common::backtrace::_print::hbfc81757b29a60f4 at src/libstd/sys_common/backtrace.rs:62
7:   0x809c1c - std::sys_common::backtrace::print::h8bc349c117633211 at src/libstd/sys_common/backtrace.rs:49
8:   0x809c1c - std::panicking::default_hook::{{closure}}::h82c47dcbb8503118 at src/libstd/panicking.rs:204
9:   0x809890 - std::panicking::default_hook::hc04800be16711189 at src/libstd/panicking.rs:224
10:   0x80a2cc - std::panicking::rust_panic_with_hook::h2413ff75e44cd581 at src/libstd/panicking.rs:472
11:   0x809ee0 - rust_begin_unwind at src/libstd/panicking.rs:380
12:   0x820df0 - core::panicking::panic_fmt::h2c56481ca2ee7c2f at src/libcore/panicking.rs:85
13:   0x820d4c - core::panicking::panic::h858859fb9e6c3f07 at src/libcore/panicking.rs:52
14:   0x4b4b94 - satnogs_monitor::ui::render_spectrum_plot::hfead694c43b2bc17 
15:   0x4765e0 - tui::terminal::Terminal<B>::draw::h407c6b0c5335bf17
16:   0x4b25e4 - satnogs_monitor::ui::Ui::draw::h2bb03ca72e177569
17:   0x4b3988 - satnogs_monitor::ui::Ui::run::hd21c6df18ba5ce3a
18:   0x460a8c - satnogs_monitor::run::h61a99fa7b0961ee2
19:   0x45cf6c - satnogs_monitor::main::h5b55de0c01d8b246
20:   0x473710 - std::rt::lang_start::{{closure}}::hb5ce5735857dd23e
21:   0x809d90 - std::rt::lang_start_internal::{{closure}}::habef164db2ffdce8 at src/libstd/rt.rs:52
22:   0x809d90 - std::panicking::try::do_call::ha24ef807c5f4c553 at src/libstd/panicking.rs:305
23:   0x80e458 - __rust_maybe_catch_panicat src/libpanic_unwind/lib.rs:86
24:   0x80a7f0 - std::panicking::try::hbf32b3a5688b523f at src/libstd/panicking.rs:281
25:   0x80a7f0 - std::panic::catch_unwind::haaedb267e2d3c6e6 at src/libstd/panic.rs:394
26:   0x80a7f0 - std::rt::lang_start_internal::hdc1e9a94360026b9 at src/libstd/rt.rs:51 
27:   0x46197c - main
28: 0xb6dc2718 - __libc_start_main

**Comment by wose** [comment]
Thank you, I'll take a look within the next 2 days. 

In the meantime you can disable the spectrum plot and it may work. There is also a known panic when your terminal is too small (#7), so increasing the size (chars per line and row) may also work.

**Comment by wose** [comment]
This should fix the panic. However, this is strange since it seems the spectrum data is present but not the frequencies (first bytes of the waterfall data file). I wasn't able to reproduce it on my side. Can you test the last commit?

What satnogs-client version are you using? Are you on armhf or aarch64?

**Comment by arope99** [comment]
Yes, it stops  the panic, but it also not showing the spectrum or waterfall, as there are no frequencies information.

I'm using satnogs-client 1.1.2 and gr-satnogs 2.0.1-1  on armhf erpi4. The frequencies information are not on the first bytes of the waterfall files.

You can see the format as shown in the following codes taken from 
https://gitlab.com/librespacefoundation/satnogs/gr-satnogs/-/blob/v2.0.1/lib/waterfall_sink_impl.cc

/*
   * Write the header. Make a dummy serialization to avoid padding and
   * alignment issues
   */
  d_fos.write(h.start_time, 32);
  d_fos.write((char *)&h.fft_size, sizeof(uint32_t));
  d_fos.write((char *)&h.samp_rate, sizeof(uint32_t));
  d_fos.write((char *)&h.nfft_per_row, sizeof(uint32_t));
  d_fos.write((char *)&h.center_freq, sizeof(float));
  d_fos.write((char *)&h.endianness, sizeof(uint32_t));

The are 32  bytes of time stamp now before the freq info.

Can you make changes to the satnogs-monitor to reflect the new format?

Thanks.

**Comment by wose** [comment]
@arope99 Thx! I missed that the change of the waterfall format was released. I'll probably have some time during the next days to work on this. This will also result in a new major release and an announcement post on the SatNOGS forum.

**Comment by arope99** [comment]
@wose Thank you. Appreciate your help!

**Comment by matburnham** [comment]
I was about to hit _Submit new issue_ to raise the following, but then saw it looks like this already covers it. Looks like it's the same problem and fixed in HEAD.

----

I ran `satnogs-monitor -l 1450 --data-path /tmp/.satnogs/data/ --waterfall --spectrum` (version  0.3.1) to see if I could watch an upcoming pass. Three seconds into the pass I got an error:

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:378:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

![image](https://user-images.githubusercontent.com/1058493/79081774-4ba4cc80-7d18-11ea-9f1b-1d431dafb0ad.png)

By the time I'd taken a copy of the error and a screenshot the pass had gone. 😦

I've added the suggested backtrace environment variable in case it happens again. It could be a while though. I wonder if there's any way to force `satnogs-client` to 'receive now' regardless of a valid satellite pass?
